### PR TITLE
Reimplement input sourcemap merging using range matching instead of closest-position matching

### LINKED ISSUE #7632: Merging inputSourceMaps may result in off-by-one lines
I'll be looking into this, but filing for tracking. Came across this with Vue's default quickstart application.

```

var vueOut = JSON.parse(String.raw`
{
  "version": 3,
  "sources": [
    "HelloWorld.vue"
  ],
  "names": [],
  "mappings": ";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAsFA;AACA;AACA;AACA;AACA;AACA;AACA;AACA",
  "sourceRoot": "src/components",
  "sourcesContent": [
    "<template>\n  <div class=\"hello\">\n    <h1>{{ msg }}</h1>\n    <h2>Essential Links</h2>\n    <ul>\n      <li>\n        <a\n          href=\"https://vuejs.org\"\n          target=\"_blank\"\n        >\n          Core Docs\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://forum.vuejs.org\"\n          target=\"_blank\"\n        >\n          Forum\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://chat.vuejs.org\"\n          target=\"_blank\"\n        >\n          Community Chat\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://twitter.com/vuejs\"\n          target=\"_blank\"\n        >\n          Twitter\n        </a>\n      </li>\n      <br>\n      <li>\n        <a\n          href=\"http://vuejs-templates.github.io/webpack/\"\n          target=\"_blank\"\n        >\n          Docs for This Template\n        </a>\n      </li>\n    </ul>\n    <h2>Ecosystem</h2>\n    <ul>\n      <li>\n        <a\n          href=\"http://router.vuejs.org/\"\n          target=\"_blank\"\n        >\n          vue-router\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"http://vuex.vuejs.org/\"\n          target=\"_blank\"\n        >\n          vuex\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"http://vue-loader.vuejs.org/\"\n          target=\"_blank\"\n        >\n          vue-loader\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://github.com/vuejs/awesome-vue\"\n          target=\"_blank\"\n        >\n          awesome-vue\n        </a>\n      </li>\n    </ul>\n  </div>\n</template>\n\n<script>\nexport default {\n  name: 'HelloWorld',\n  data () {\n    return {\n      msg: 'Welcome to Your Vue.js App'\n    }\n  }\n}\n</script>\n\n<!-- Add \"scoped\" attribute to limit CSS to this component only -->\n<style scoped>\nh1, h2 {\n  font-weight: normal;\n}\nul {\n  list-style-type: none;\n  padding: 0;\n}\nli {\n  display: inline-block;\n  margin: 0 10px;\n}\na {\n  color: #42b983;\n}\n</style>\n"
  ]
}
`);

var babelOut = JSON.parse(String.raw`
{
  "version": 3,
  "sources": [
    "src/components/HelloWorld.vue"
  ],
  "names": [
    "name",
    "data",
    "msg"
  ],
  "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;AAEA,eAAe;AACbA,QAAM,YADO;AAEbC,MAFa,kBAEL;AACN,WAAO;AACLC,WAAK;AADA,KAAP;AAGD;AANY,CAAf",
  "file": "HelloWorld.vue",
  "sourceRoot": "/Users/thesmyth/projects/debugger-examples/examples/vue",
  "sourcesContent": [
    "//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n//\n\nexport default {\n  name: 'HelloWorld',\n  data () {\n    return {\n      msg: 'Welcome to Your Vue.js App'\n    }\n  }\n}\n"
  ]
}
`);
```

With this, running
```
require("./packages/babel-core/lib/transformation/file/generate").mergeSourceMap(vueOut, babelOut)
```

results in
```
{
  "version": 3,
  "sources": [
    "HelloWorld.vue"
  ],
  "names": [],
  "mappings": ";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAsFA;QAEA;wBACA;;WACA,AACA,AACA,AACA,AANA",
  "sourceRoot": "src/components",
  "sourcesContent": [
    "<template>\n  <div class=\"hello\">\n    <h1>{{ msg }}</h1>\n    <h2>Essential Links</h2>\n    <ul>\n      <li>\n        <a\n          href=\"https://vuejs.org\"\n          target=\"_blank\"\n        >\n          Core Docs\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://forum.vuejs.org\"\n          target=\"_blank\"\n        >\n          Forum\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://chat.vuejs.org\"\n          target=\"_blank\"\n        >\n          Community Chat\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://twitter.com/vuejs\"\n          target=\"_blank\"\n        >\n          Twitter\n        </a>\n      </li>\n      <br>\n      <li>\n        <a\n          href=\"http://vuejs-templates.github.io/webpack/\"\n          target=\"_blank\"\n        >\n          Docs for This Template\n        </a>\n      </li>\n    </ul>\n    <h2>Ecosystem</h2>\n    <ul>\n      <li>\n        <a\n          href=\"http://router.vuejs.org/\"\n          target=\"_blank\"\n        >\n          vue-router\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"http://vuex.vuejs.org/\"\n          target=\"_blank\"\n        >\n          vuex\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"http://vue-loader.vuejs.org/\"\n          target=\"_blank\"\n        >\n          vue-loader\n        </a>\n      </li>\n      <li>\n        <a\n          href=\"https://github.com/vuejs/awesome-vue\"\n          target=\"_blank\"\n        >\n          awesome-vue\n        </a>\n      </li>\n    </ul>\n  </div>\n</template>\n\n<script>\nexport default {\n  name: 'HelloWorld',\n  data () {\n    return {\n      msg: 'Welcome to Your Vue.js App'\n    }\n  }\n}\n</script>\n\n<!-- Add \"scoped\" attribute to limit CSS to this component only -->\n<style scoped>\nh1, h2 {\n  font-weight: normal;\n}\nul {\n  list-style-type: none;\n  padding: 0;\n}\nli {\n  display: inline-block;\n  margin: 0 10px;\n}\na {\n  color: #42b983;\n}\n</style>\n"
  ]
}
```
which when converted to readable mappings is
```
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 87,
  generatedColumn: 0,
  originalLine: 87,
  originalColumn: 0,
  name: null }
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 88,
  generatedColumn: 8,
  originalLine: 89,
  originalColumn: 0,
  name: null }
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 89,
  generatedColumn: 24,
  originalLine: 90,
  originalColumn: 0,
  name: null }
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 91,
  generatedColumn: 11,
  originalLine: 88,
  originalColumn: 0,
  name: null }
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 91,
  generatedColumn: 11,
  originalLine: 91,
  originalColumn: 0,
  name: null }
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 91,
  generatedColumn: 11,
  originalLine: 92,
  originalColumn: 0,
  name: null }
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 91,
  generatedColumn: 11,
  originalLine: 93,
  originalColumn: 0,
  name: null }
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 91,
  generatedColumn: 11,
  originalLine: 94,
  originalColumn: 0,
  name: null }
```

The mappings for Vue's output are neat line to line mappings, and the output from Babel itself seems alright from what I can tell, but things get complicated when they are merged somewhere, resulting in
```
{ source: 'src/components/HelloWorld.vue',
  generatedLine: 88,
  generatedColumn: 8,
  originalLine: 89,
  originalColumn: 0,
  name: null }
```
above for instance, where I'd expect the generated and original line to be the same.

### PR DESCRIPTION
**Pull request by loganfsmyth**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            |  Fixes #7632
| Patch: Bug Fix?          | Y
| Major: Breaking Change?  | Not unless things were relying on our previous awful maps
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR         | 
| Any Dependency Changes?  |
| License                  | MIT

This PR ended up bigger than I'd originally expected, so sorry about that. The core of it comes down to the fact that just calling `generatedPositionFor` is an extremely bad/inaccurate way to merge two sourcemaps.

The core of this PR is making full use of the fact that source mappings are generally processed as _ranges_ rather than individual locations. The new code walks from the original input mappings down to the actual mappings generated by Babel, but does so taking ranges into account. Rather than blindly calling `generatedPositionFor`, we explicitly search for the set of ranges that overlap.

This fix is driven by my work on FF in https://github.com/devtools-html/debugger.html/issues/5561

### CODE REVIEW

File: packages/babel-core/src/transformation/file/merge-map.js [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+      // skip generated locations once we've seen them the first time.
+      const key = JSON.stringify(item);
+      if (mappedGeneratedLocations.has(key)) return;
+      mappedGeneratedLocations.add(key);
+
+      mergedGenerator.addMapping({
+        source: source.path,
+        original: {
+          line: original.line,
+          column: original.column,
```
[the remark this answers is not in the document]
Comment (loganfsmyth) [reply packages/babel-core/src/transformation/file/merge-map.js]: I don't follow that logic. If the input sourcemap had a range over a set of locations, isn't it reasonable to expect exactly the same range over the original files in the output map?

[the remark this answers is not in the document]
Comment (gzzhanghao) [reply packages/babel-core/src/transformation/file/merge-map.js]: I'm not sure about that, in vue-loader's case column info are missing with `original.column`.

`original.column + item.original.column - generated.column` is based on an assumption that the mapped range will be exactly same as the original one. There may be exceptions, so I'm not sure if that will lead to any problems.

[the remark this answers is not in the document]
Comment (loganfsmyth) [reply packages/babel-core/src/transformation/file/merge-map.js]: > is based on an assumption that the mapped range will be exactly same as the original one.

That is my primary concern. I don't think that's a safe assumption to make for merge logic that should ideally be fairly generic. Any kind of math trying to calculate column numbers would be making assumptions about the structure of the code being mapped. I don't think I'm personally comfortable with doing that.

**Comment by loganfsmyth** [comment]
Fixed a few small issues with this. I'm probably going to merge it after the tests pass, since it seems like realistically no-one has the background to review it...

**Pull request merged**